usr/sbin/p11sak/p11sak.c

```diff
@@ -1171,41 +1335,56 @@ static CK_RV sec_key_print_attributes(CK_SESSION_HANDLE session,
-    CK_ULONG bs = sizeof(CK_BBOOL);
 
-    CK_ATTRIBUTE bool_tmplt[] = { { CKA_TOKEN, &a_token, bs }, { CKA_PRIVATE,
-            &a_private, bs }, { CKA_MODIFIABLE, &a_modifiable, bs }, {
-    CKA_DERIVE, &a_derive, bs }, { CKA_LOCAL, &a_local, bs }, {
-    CKA_SENSITIVE, &a_sensitive, bs }, { CKA_ENCRYPT, &a_encrypt, bs }, {
-            CKA_DECRYPT, &a_decrypt, bs }, { CKA_SIGN, &a_sign, bs }, {
-    CKA_VERIFY, &a_verify, bs }, { CKA_WRAP, &a_wrap, bs }, {
-    CKA_UNWRAP, &a_unwrap, bs }, { CKA_EXTRACTABLE, &a_extractable, bs }, {
-            CKA_ALWAYS_SENSITIVE, &a_always_sensitive, bs }, {
-            CKA_NEVER_EXTRACTABLE, &a_never_extractable, bs }, };
+    CK_ATTRIBUTE bool_tmplt[] = { 
+        { CKA_TOKEN, &a_token, sizeof(a_token) }, 
+        { CKA_PRIVATE, &a_private, sizeof(a_private) }, 
+        { CKA_MODIFIABLE, &a_modifiable, sizeof(a_modifiable) }, 
+        { CKA_DERIVE, &a_derive, sizeof(a_derive) }, 
+        { CKA_LOCAL, &a_local, sizeof(a_local) }, 
+        { CKA_SENSITIVE, &a_sensitive, sizeof(a_sensitive) }, 
+        { CKA_ENCRYPT, &a_encrypt, sizeof(a_encrypt) }, 
+        { CKA_DECRYPT, &a_decrypt, sizeof(a_decrypt) }, 
+        { CKA_SIGN, &a_sign, sizeof(a_sign) }, 
+        { CKA_VERIFY, &a_verify, sizeof(a_verify) }, 
+        { CKA_WRAP, &a_wrap, sizeof(a_wrap) }, 
+        { CKA_UNWRAP, &a_unwrap, sizeof(a_unwrap) }, 
+        { CKA_EXTRACTABLE, &a_extractable, sizeof(a_extractable) }, 
+        { CKA_ALWAYS_SENSITIVE, &a_always_sensitive, sizeof(a_always_sensitive) }, 
+        { CKA_NEVER_EXTRACTABLE, &a_never_extractable, sizeof(a_never_extractable) } };
     CK_ULONG count = sizeof(bool_tmplt) / sizeof(CK_ATTRIBUTE);
 
     rc = funcs->C_GetAttributeValue(session, hkey, bool_tmplt, count);
     if (rc != CKR_OK) {
-        printf("Attribute retrieval failed (error code 0x%lX: %s)\n", rc,
+        fprintf(stderr, "Attribute retrieval failed (error code 0x%lX: %s)\n", rc,
                 p11_get_ckr(rc));
         return rc;
     }
 
     if (long_print) {
         for (i = 0; i < (int) count; i++) {
             if (bool_tmplt[i].ulValueLen != sizeof(CK_BBOOL)) {
-                printf(" Error in retrieving Attribute %s\n",
+                fprintf(stderr, " Error in retrieving Attribute %s\n",
                         CKA2a(bool_tmplt[i].type));
             } else {
                 printf("          %s: %s\n", CKA2a(bool_tmplt[i].type),
                         CK_BBOOL2a(*(CK_BBOOL*) bool_tmplt[i].pValue));
             }
         }
-        printf("|\n");
+        rc = print_vendor(session, hkey, long_print);
+        if (rc != CKR_OK) {
+            return rc;
+        }
     } else {
         printf(" |");
         for (i = 2; i < KEY_MAX_BOOL_ATTR_COUNT; i++)
             short_print(i, bool_tmplt, kt_SECRET);
+        rc = print_vendor(session, hkey, long_print);
+        if (rc != CKR_OK) {
+            fprintf(stderr, "Attribute retrieval failed (error code 0x%lX: %s)\n", rc,
```

Comment:
Why is here an error message but not in Line 1376?  Either put all error messages in `print_vendor` which already contains error messages for almost all non-`CKR_OK` returns, or put them consistently in the caller.

Reply:
Done